### PR TITLE
JW-1068: implement GUIDs for vifs

### PR DESCRIPTION
**Pull request by sodar**

### CODE REVIEW

File: src/vnsw/agent/vrouter/ksync/interface_ksync.cc [hunk 1]
```diff
@@ -732,6 +740,16 @@ int InterfaceKSyncEntry::Encode(sandesh_op::type op, char *buf, int buf_len) {
     encoder.set_vifr_ip(ip_);
     encoder.set_vifr_nh_id(flow_key_nh_id_);
 
+    /* GUIDs are used as interface identifiers on Windows */
+    if (os_guid_) {
+        auto& os_guid = os_guid_.get();
```
Comment (jablonskim) [inline src/vnsw/agent/vrouter/ksync/interface_ksync.cc]: I think we cannot use `auto` (and other C++11 features) in linux code or system-independent code

File: src/vnsw/agent/oper/interface.cc [hunk 2]
```diff
@@ -469,7 +469,38 @@ NET_LUID GetInterfaceLuidFromName(const std::string& name, const Interface::Type
...
+        assert(0);
+    }
+
+    Interface::IfGuid result;
+    memcpy(&result, &intf_guid, sizeof(intf_guid));
+
+    return result;
+}
+
+NET_LUID GetInterfaceLuidFromGuid(const Interface::IfGuid intf_guid) {
```
Comment (jablonskim) [inline src/vnsw/agent/oper/interface.cc]: I think it would be better to pass GUID by const reference instead of by const value

**Review by jablonskim — CHANGES_REQUESTED** [review]

**Review by jablonskim — APPROVED** [review]

File: src/vnsw/agent/vrouter/ksync/interface_ksync.cc [hunk 3]
```diff
@@ -116,7 +117,8 @@ InterfaceKSyncEntry::InterfaceKSyncEntry(InterfaceKSyncObject *obj,
     no_arp_(false),
     encap_type_(PhysicalInterface::ETHERNET),
     transport_(Interface::TRANSPORT_INVALID),
-    flood_unknown_unicast_(false), qos_config_(NULL) {
+    flood_unknown_unicast_(false), qos_config_(NULL),
```
Comment (mclapinski) [inline src/vnsw/agent/vrouter/ksync/interface_ksync.cc]: You could split it into 2 lines

File: src/vnsw/agent/oper/interface.h [hunk 4]
```diff
@@ -175,9 +179,8 @@ class Interface : AgentRefCount<Interface>, public AgentOperDBEntry {
     Transport transport_;
     AgentQosConfigConstRef qos_config_;
 
-#ifdef _WIN32
-    boost::optional<NET_LUID> intf_luid_;
-#endif
+    // Used on Windows as operating system identifier
```
Comment (mclapinski) [inline src/vnsw/agent/oper/interface.h]: sounds like it identifies operating system

File: src/vnsw/agent/oper/interface.cc [hunk 5]
```diff
@@ -579,15 +610,25 @@ void Interface::GetOsParams(Agent *agent) {
...
+
+        os_index_ = GetInterfaceIndexFromLuid(net_luid);
+        name_ = GetInterfaceNameFromLuid(net_luid);
+        mac_ = GetMacAddressFromLuid(net_luid);
+
+        /* We assume that interface is UP */
+        os_oper_state_ = true;
+    } else {
+        LOG(ERROR, "interface GUID not set on " << name);
+        assert(0);
```
Comment (mclapinski) [inline src/vnsw/agent/oper/interface.cc]: Now this `else` is useless. When you change the return type of GetInterfaceGuidFromLuid it probably will be usefull.

File: src/vnsw/agent/oper/interface.cc [hunk 6]
```diff
@@ -469,7 +469,38 @@ NET_LUID GetInterfaceLuidFromName(const std::string& name, const Interface::Type
...
 }
 
-NET_IFINDEX GetInterfaceIndexFromLuid(NET_LUID intf_luid) {
+Interface::IfGuid GetInterfaceGuidFromLuid(const NET_LUID intf_luid) {
+    GUID intf_guid;
+
+    NETIO_STATUS status = ConvertInterfaceLuidToGuid(&intf_luid, &intf_guid);
+    if (status != NO_ERROR) {
+        LOG(ERROR, "ERROR: on converting LUID to GUID:" << status);
+        assert(0);
```
Comment (mclapinski) [inline src/vnsw/agent/oper/interface.cc]: return null or sth

File: src/vnsw/agent/oper/interface.cc [hunk 7]
```diff
@@ -469,7 +469,38 @@ NET_LUID GetInterfaceLuidFromName(const std::string& name, const Interface::Type
...
+    NET_LUID intf_luid;
+
+    GUID win_guid;
+    assert(sizeof(win_guid) == intf_guid.size());
+    memcpy(&win_guid, &intf_guid, intf_guid.size());
+
+    NETIO_STATUS status = ConvertInterfaceGuidToLuid(&win_guid, &intf_luid);
+    if (status != NO_ERROR) {
+        LOG(ERROR, "ERROR: on converting GUID to LUID:" << status);
+        assert(0);
```
Comment (mclapinski) [inline src/vnsw/agent/oper/interface.cc]: same

**Review by mclapinski — CHANGES_REQUESTED** [review]

**Review by jablonskim — APPROVED** [review]

File: src/vnsw/agent/oper/interface.cc [hunk 8]
```diff
@@ -559,7 +559,17 @@ MacAddress GetMacAddressFromLuid(const NET_LUID intf_luid) {
...
+static std::string LuidToString(const NET_LUID intf_luid) {
+    std::stringstream ss;
+
+    ss << "<IfType = " << intf_luid.Info.IfType
+       << ", NetLuidIndex = " << intf_luid.Info.NetLuidIndex
+       << ", Value = " << intf_luid.Value << ">";
+
+    return ss.str();
 }
 #endif
```
Comment (mclapinski) [inline src/vnsw/agent/oper/interface.cc:574]: How big is this #ifdef? Don't you want to move this to another file (Windows only)?

File: src/vnsw/agent/oper/interface.cc [hunk 9]
```diff
@@ -612,24 +622,61 @@ void Interface::GetOsParams(Agent *agent) {
...
+
+    os_index_ = *os_index;
+    name_ = *if_name;
+    mac_ = *mac;
+
+    /* We assume that interface is UP */
+    os_oper_state_ = true;
+
+    return;
 #else
```
Comment (mclapinski) [inline src/vnsw/agent/oper/interface.cc]: You could also move this #ifdef to that file.

**Review by mclapinski — APPROVED** [review]

File: src/vnsw/agent/oper/interface.cc (hunk 8 above)
Comment (sodar) [reply src/vnsw/agent/oper/interface.cc:574]: I think that is a good idea. I have created a task in Jira (JW-1107) for this, so it can be done later.

**Pull request merged**